Add documentation for OpenSearch Flow OSD plugin (#9101)
opensearch-project/documentation-website: _automating-configurations/workflow-builder.md

    ![Transform response]({{site.url}}{{site.baseurl}}/images/dashboards-flow-framework/transform-response.png)
    ![Configure prompt]({{site.url}}{{site.baseurl}}/images/dashboards-flow-framework/configure-prompt.png)
12. Click **Update**, to build out the search pipeline. The **Inspect flows** section will automatically navigate to **Test flow**. From here, you may test out different queries and execute search. Variables wrapped in `{{ }}` allow for rapidly testing out different query values, without changing the base query.
    ![Test search flow]({{site.url}}{{site.baseurl}}/images/dashboards-flow-framework/search-test-flow.png)
13. View results. After clicking **Search**, view the results as either a formatted list of hits, or the raw JSON search response.
14. Tune your configurations to satisfactorily fit your use case. There are many ways this can be done:
    - Try different query parameters
    - Try different queries
    - Update existing processors under **Enhance query request** or **Enhance query response**
    - Add or remove processors under **Enhance query request** or **Enhance query response**

> 15. Export your workflow. Click on **Export** in the header to bring up the pop-up window. The displayed data represents the [Workflow Template]({{site.url}}{{site.baseurl}}/automating-configurations/workflow-templates/) containing the end-to-end configuration for building out the OpenSearch resources; in this case, the ingest pipeline, index, and search pipeline. You may download in JSON or YAML format from the button on the right-hand side.

Reviewer: Does it make sense to call out what can users do with the export workflow? Perhaps link to the workflow API on how to "Deploy"?

Author: Agreed. added

   - The large language model (LLM) to perform RAG
   ![Quick configure modal]({{site.url}}{{site.baseurl}}/images/dashboards-flow-framework/quick-configure-modal.png)
   
   For additional options, such as the text field and vector field names that will be persisted in the index, select **Optional configuration**. You can update these settings at any time.
4. Select **Create** to prepopulate the configuration and automatically navigate to the **Workflow Details** page, where you can configure your ingest flow.
5. To provide sample data, select **Import**. You can either manually enter data, upload a local `.json` file, or retrieve sample documents from an existing index, as shown in the following image. 
   ![Import data modal]({{site.url}}{{site.baseurl}}/images/dashboards-flow-framework/import-data-modal.png)

   The form expects an array of JSON objects, with each object representing a standalone [document]({{site.url}}{{site.baseurl}}/getting-started/intro/#document). This process is similar to the [bulk ingest operation]({{site.url}}{{site.baseurl}}/getting-started/ingest-data/#bulk-indexing). When you're finished, select **Save**.

> 6. Under **Transform data**, select **ML Inference Processor**. This section will be pre-populated with configurations to map data _to_ the expected model input and _from_ the expected model output, as shown in the following image. This example, under **Inputs**, maps the target document field to the model input field, generating vector embeddings for that field. Under **Outputs**, it maps the model output field to a new field, which will be persisted in the index.

Reviewer 2: ```suggestion
6. Under **Transform data**, select **ML Inference Processor**. This section will be prepopulated with configurations used to map data _to_ the expected model input and _from_ the expected model output, as shown in the following image. This example, under **Inputs**, maps the target document field to the model input field, generating vector embeddings for that field. Under **Outputs**, it maps the model output field to a new field, which will be persisted in the index.
```

   ![Transform data]({{site.url}}{{site.baseurl}}/images/dashboards-flow-framework/transform-data.png)

>    For more information transformation types that accommodate complex data schemas and model interfaces, see [Advanced data transformations](#advanced-data-transformations).

Reviewer 2: ```suggestion
   For more information about transformation types that accommodate complex data schemas and model interfaces, see [Advanced data transformations](#advanced-data-transformations).
```

> 7. Under **Ingest data**, select **Advanced settings**. This section will be pre-populated with the index configurations required for this use case, such as the `knn_vector` field mapping for the vector embedding field and the index setting `index.knn: true`, as shown in the following image. You can adjust these settings as needed.

Reviewer 2: ```suggestion
7. Under **Ingest data**, select **Advanced settings**. This section will be prepopulated with the index configurations required for this use case, such as the `knn_vector` field mapping for the vector embedding field and the index setting `index.knn: true`, as shown in the following image. You can adjust these settings as needed.
```

>    ![Ingest data]({{site.url}}{{site.baseurl}}/images/dashboards-flow-framework/ingest-data.png)

Reviewer 2: Line 72: Should `vector embedding` be in code font?

Reviewer 3: No because this is an arbitrary field.

> 8. Select **Update** in the bottom bar to build the configured ingest pipeline and index, and ingest the provided sample data. Afterward, go to **Test flow** under **Inspect flows** to search the newly created index and verify that the transformed documents are as expected. In this example, verify that the vector embeddings were generated for each ingested document, as shown in the following image.

Reviewer 2: ```suggestion
8. Select **Update** in the bottom bar to build the configured ingest pipeline and index and ingest the provided sample data. After this, go to **Test flow** under **Inspect flows** to search the newly created index and verify that the transformed documents are as expected. In this example, verify that the vector embeddings were generated for each ingested document, as shown in the following image.
```

>    ![Test ingest flow]({{site.url}}{{site.baseurl}}/images/dashboards-flow-framework/ingest-test-flow.png)

Reviewer 2: Line 74: Please rewrite "in the bottom bar". Do we mean something like "on the bottom menu bar"? In the second sentence, please rewrite "are as expected". What do we mean by "are"? "appear as expected"?

> 9. To configure your search flow, select **Next: Search flow**. Like the ingest flow, the form will be pre-populated with initial configuration settings.

Reviewer 2: ```suggestion
9. To configure your search flow, select **Next: Search flow**. Like the ingest flow, the form will be prepopulated with initial configuration settings.
```

10. Under **Transform query**, select **ML Inference Processor**, as shown in the following image. This processor parses the inputs in the search query for which you want to generate vector embeddings. In this example, it passes the value from `query.term.item_text.value` to the embedding model.<br>
    ![Transform query]({{site.url}}{{site.baseurl}}/images/dashboards-flow-framework/transform-query.png)

>     The processor also performs a query rewrite to generate a `knn` query using the vector embedding produced by the model. Select **Rewrite** to view its details, as shown in the following image. This approach abstracts complex query details, providing a simple query interface that uses search pipelines to perform advanced query generation. 

Reviewer 2: Should the singular instance of "embedding" be plural?

>     The processor also performs a query rewrite to generate a `knn` query using the vector embedding produced by the model. Select **Rewrite** to view its details, as shown in the following image. This approach abstracts complex query details, providing a simple query interface that uses search pipelines to perform advanced query generation. 

Reviewer 3: ```suggestion
    The processor also performs a query rewrite to generate a `knn` query using the vector embeddings produced by the model. Select **Rewrite** to view its details, as shown in the following image. This approach abstracts complex query details, providing a simple query interface that uses search pipelines to perform advanced query generation. 
```

    ![Rewrite query]({{site.url}}{{site.baseurl}}/images/dashboards-flow-framework/rewrite-query.png)

> 11. Under **Transform response**, select **ML Inference Processor**, as shown in the following image. This is where the Claude LLM is used to process the returned results and generate a human-readable response. 

Reviewer 2: ```suggestion
11. Under **Transform response**, select **ML Inference Processor**, as shown in the following image. The Claude LLM is used to process the returned results and generate a human-readable response. 
```

   ![Transform response]({{site.url}}{{site.baseurl}}/images/dashboards-flow-framework/transform-response.png)<br>
   Under **Inputs**, select the pencil icon next to the `prompt` entry. A pop-up will appear with a preconfigured prompt template designed to summarize the returned documents, as shown in the following image. You can adjust this template as needed, and several presets are available as starting points. You can also add, update, or remove the **Input variables**, which include data from the returned documents that you want to dynamically inject as contextual information to the LLM. The default option collects all `item_text` data and summarizes the results. Select **Save** to apply your changes.

>     ![Configure prompt]({{site.url}}{{site.baseurl}}/images/dashboards-flow-framework/configure-prompt.png)

Reviewer 3: Rewritten

>     ![Configure prompt]({{site.url}}{{site.baseurl}}/images/dashboards-flow-framework/configure-prompt.png)

Reviewer 2: Works for me 👍

> 12. Select **Update** to build the search pipeline. The **Inspect flows** section will automatically navigate to **Test flow**, where you can test different queries and run searches, as shown in the following image. You can use variables wrapped in {% raw %}`{{ }}`{% endraw %} to quickly test different query values without modifying the base query.

Reviewer 3: ```suggestion
12. Select **Update** to build the search pipeline. The **Inspect flows** section will automatically navigate to the **Test flow** section, where you can test different queries and run searches, as shown in the following image. You can use variables wrapped in {% raw %}`{{ }}`{% endraw %} to quickly test different query values without modifying the base query.
```

    ![Test search flow]({{site.url}}{{site.baseurl}}/images/dashboards-flow-framework/search-test-flow.png)

> 13. To view results, select **Search**. You can view the results as either a formatted list of hits or as the raw JSON search response.

Reviewer 3: ```suggestion
13. To view the search results, select **Search**. You can view the results as either a formatted list of hits or as the raw JSON search response.
```

14. Adjust your configurations to better suit your use case. You can do this in several ways:
   - Experiment with different query parameters.
   - Try different queries.
   - Modify existing processors under **Enhance query request** or **Enhance query response**.
   - Add or remove processors under **Enhance query request** or **Enhance query response**.
15. Export your workflow by selecting **Export** in the header to open the pop-up window. The displayed data represents the [Workflow Template]({{site.url}}{{site.baseurl}}/automating-configurations/workflow-templates/), which contains the full configuration for the OpenSearch resources you've just created, including the ingest pipeline, index, and search pipeline. You can download the template in JSON or YAML format by clicking the button on the right-hand side. To build identical resources in other OpenSearch clusters, use the [Provision Workflow API]({{site.url}}{{site.baseurl}}/automating-configurations/api/provision-workflow/).

## Advanced data transformations

ML inference processors provide several flexible ways to transform input data _to_ the model input(s) and _from_ the model output(s).

> In **Inputs**, you can configure the parameters passed _to_ the model. Each input parameter can be one of four transformation types:

Reviewer 2: ```suggestion
In **Inputs**, you can configure the parameters passed _to_ the model. There are four input parameter transformation types:
```

> 1. **Data field**: Use an existing field from your data as the model input.

Reviewer 2: ```suggestion
1. **Data field**: Uses an existing data field as the model input.
```

> 2. **JSONPath expression**: Extract data from a JSON structure and map the extracted data to the model input field using [JSONPath](https://en.wikipedia.org/wiki/JSONPath).

Reviewer 2: ```suggestion
2. **JSONPath expression**: Extracts data from a JSON structure and maps the extracted data to the model input field using [JSONPath](https://en.wikipedia.org/wiki/JSONPath).
```

> 3. **Prompt**: Use a constant value that can include dynamic variables. This combines elements of both the `Custom string` transformation and the `Data field` and `JSONPath expression` transformations, making it especially useful for building prompts for LLMs.

Reviewer 2: ```suggestion
3. **Prompt**: Uses a constant value that can include dynamic variables. This combines elements of both the `Custom string` transformation and the `Data field` and `JSONPath expression` transformations, making it especially useful for building prompts for LLMs.
```

> 4. **Custom string**: Use a constant string value.

Reviewer 2: ```suggestion
4. **Custom string**: Uses a constant string value.
```

> In **Outputs**, you can configure the values passed _from_ the model. Each output parameter can be one of three transformation types:

Reviewer 2: ```suggestion
In **Outputs**, you can configure the values passed _from_ the model. There are three output parameter transformation types:
```

> 1. **Data field**: Copy the model output into a new document field.

Reviewer 2: ```suggestion
1. **Data field**: Copies the model output into a new document field.
```

> 2. **JSONPath Expression**: Extract data from a JSON structure and map the extracted data to one or more new document fields using [JSONPath](https://en.wikipedia.org/wiki/JSONPath).

Reviewer 2: ```suggestion
2. **JSONPath Expression**: Extracts data from a JSON structure and maps the extracted data to one or more new document fields using [JSONPath](https://en.wikipedia.org/wiki/JSONPath).
```

> 3. **No transformation**: Leave the model output field as is, preserving both its name and value.

Reviewer 2: ```suggestion
3. **No transformation**: Does not transform the model output field, preserving both its name and value.
```

## Next steps

> - For suggested models and model interfaces to use within OpenSearch Flow, see [Models](https://github.com/opensearch-project/dashboards-flow-framework/blob/main/documentation/models.md).

Reviewer 2: ```suggestion
- For models and model interfaces recommended for use with OpenSearch Flow, see [Models](https://github.com/opensearch-project/dashboards-flow-framework/blob/main/documentation/models.md).
```

> - For additional tutorials on using OpenSearch Flow to build different AI/ML use cases, see [Tutorial](https://github.com/opensearch-project/dashboards-flow-framework/blob/main/documentation/tutorial.md).

Reviewer 2: ```suggestion
- For a tutorial demonstrating how to use OpenSearch Flow to build different AI/ML use cases, see [Tutorial](https://github.com/opensearch-project/dashboards-flow-framework/blob/main/documentation/tutorial.md).
```

> 

Reviewer 2: The note I deleted would be more appropriate to a blog post as opposed to the documentation.

> _Notice a missing ingest or search processor you'd like to see in the plugin? Consider opening an [issue](https://github.com/opensearch-project/dashboards-flow-framework/issues) or contributing by opening a [pull request](https://github.com/opensearch-project/dashboards-flow-framework/pulls)._

Reviewer 2: ```suggestion
```

> {: .note}

Reviewer 2: ```suggestion
```

---
layout: default
title: Build workflows
nav_order: 15
---

# Build workflows with OpenSearch Flow

The OpenSearch Flow plugin in OpenSearch Dashboards lets you iteratively build and test use cases leveraging ingest and search pipelines. It aims to simplify the complexity around building different AI/ML use cases using ML inference processors, such as vector search and retrieval-augmented generation (RAG). Behind the scenes, the plugin uses the [Flow Framework OpenSearch Plugin]({{site.url}}{{site.baseurl}}/automating-configurations/index/) for resource management. Once you have built out a use case to your satisfaction, you can export the [Workflow Template]({{site.url}}{{site.baseurl}}/automating-configurations/workflow-templates/) to re-create identical resources across multiple clusters.

## Background: Ingest pipelines, search pipelines, and ML models

> [Ingest pipelines]({{site.url}}{{site.baseurl}}/ingest-pipelines/) and [search pipelines]({{site.url}}{{site.baseurl}}/search-plugins/search-pipelines/index/) are used for configuring data transforms at different stages of ingest and search operations using individual processors, all run within OpenSearch. An _ingest pipeline_ is composed of a sequence of _ingest processors_, while a _search pipeline_ is composed of a sequence of _search request processors_ and/or _search response processors_. You can stitch together processors and build custom pipelines to fit your data processing needs.

Reviewer 4: can we add links to search request processors and search response processors. ex: https://opensearch.org/docs/latest/search-plugins/search-pipelines/search-processors/#search-request-processors. Likewise for ingest processors as well?

Author: updated

Reviewer: +1, do we want to list out and describe the available processors for ingest, search request and search response? Does ingest flow support every processors in the ingest pipeline? If not, it may be helpful to distinguish this.

Author: We currently only support a small subset of processors (both ingest & search) that we've found most important for building out these flows, without overwhelming options. I'm weary to add explicit lists here as it may change often, and don't want the documentation to become out of date, as well as the fact that we will have different list of supported processors based on the datasource version. This is part of the reason I've added the call to action at the bottom to request for more processors to be added, should users find them helpful. WDYT?

As we gather more user feedback, the list will inevitably grow. In the future, we may be able to point to a github file containing the supported list of processors by version, in a clean manner. This can allow supporting a moving target with a static link, like I've done with the models link pointing to https://github.com/opensearch-project/dashboards-flow-framework/blob/main/documentation/models.md

In general, these pipelines allow for modifying data at 3 different stages.

1. **Ingest**: transform your documents before they are ingested in an index.
2. **Search request**: transform the search request before executing search against an index.
3. **Search response**: transform the search response (and documents) after search is executed, but before the response is returned to the user.

The [ML Commons OpenSearch Plugin]({{site.url}}{{site.baseurl}}/ml-commons-plugin/) lets you [integrate models hosted on third-party platforms]({{site.url}}{{site.baseurl}}/ml-commons-plugin/remote-models/index/), and use their inference capabilities directly within OpenSearch. Both ingest and search pipelines offer [ML inference processors]({{site.url}}{{site.baseurl}}/ingest-pipelines/processors/ml-inference/), allowing you to leverage remote model inference within your pipelines at both ingest and search time. This plugin helps in building, testing, and visualizing ingest and search pipelines for your particular use case, including the use of ML inference processors.

> 

Reviewer 4: instead of "This plugin helps in building, testing, and visualizing ingest and search pipelines for your particular use case, including the use of ML inference processors.", can we say "OpenSearch Flow plugin helps in building, testing, and visualizing ingest and search pipelines for your particular use case, including the use of ML inference processors."

Author: updated

![Transform data]({{site.url}}{{site.baseurl}}/images/dashboards-flow-framework/transform-data.png)

7. Click the **Advanced settings** accordion under **Ingest data** to see its details. This will also be pre-populated with the index configurations required for this particular use case, such as the `knn_vector` field mapping for the vector embedding field, and the index setting `index.knn: true`. You may tune these as you like.

![Ingest data]({{site.url}}{{site.baseurl}}/images/dashboards-flow-framework/ingest-data.png)

8. Click **Update** in the bottom bar to build out the configured ingest pipeline and index, as well as ingest the provided sample data. You may then navigate to **Test flow** under **Inspect flows** to search against the newly-created index and verify the transformed documents look as expected. For this example, you would want to ensure that the vector embeddings were generated for each ingested document.

![Test ingest flow]({{site.url}}{{site.baseurl}}/images/dashboards-flow-framework/ingest-test-flow.png)

> 9. Click **Next: Search pipeline** to begin configuring your search flow. Similar to ingest, the form will be pre-populated with some initial configuration.

Reviewer: Should the button be `Next: Search flow`?

10. Click the **ML Inference Processor** accordion under **Transform query** to see its details. The purpose of this processor is to parse out inputs in the search query that you want to generate vector embeddings for. In this particular example, it is passing the value in `query.term.item_text.value` to the embedding model. Additionally, it is performing a query rewrite, to generate a `knn` query using the produced vector embedding from the model. Click on **Rewrite** to see its details. This is one way to abstract out complex query details, and provide a simple query interface, leveraging search pipelines for performing the advanced query generation.

![Transform query]({{site.url}}{{site.baseurl}}/images/dashboards-flow-framework/transform-query.png)

![Rewrite query]({{site.url}}{{site.baseurl}}/images/dashboards-flow-framework/rewrite-query.png)

> 11. Click the **ML Inference Processor** accordion under **Transform response** to see its details. This is where the Claude LLM is leveraged to collect the returned results, and return some human-readable response. Under **Inputs**, click the pencil icon on the right-hand side of the `prompt` entry. This will display a pop-up with a preconfigured prompt template, aimed at summarizing the returned documents. You may tune this as you like, and there are several presets provided as starting points. You may add/update/remove the **Input variables**, which contain the data in the returned documents you want to dynamically inject as contextual information to the LLM. The preconfigured option shows collecting all of the `item_text` data, and summarizing the results. Click **Save** or **Cancel** to exit the pop-up window.

Reviewer: > Click **Save** or **Cancel** to exit the pop-up window.

Perhaps we can just say `Click **Save** to apply changes.`?

Author: updated
